01-messaging.md

```diff
@@ -153,6 +156,73 @@ diagnosis, as it indicates that one peer has a bug.
 It may be wise not to distinguish errors in production settings, lest
 it leak information, thus the optional data field.
 
+## Control Messages
+
+### The `ping` message
+
+In order to allow for the existence of very long-lived TCP connections, at
+times it may be required that both ends keep alive the TCP connection at the
+application level.
+
+1. type: 18 (`ping`)
+2. data: 
+    * [2:num_pong_bytes]
+
+#### Requirements
+
+A node receiving a `ping` message MUST respond with a `pong` message that
+contains the specified `num_pong_bytes` number of bytes within it's data
+payload. These `ping` messages may be sent any time at the sender's discretion,
+however `ping` messages MUST NOT be sent with a frequency lower than one every
+30 seconds.
+
+### The `pong` message
+
+The `pong` message is to be sent whenever a `ping` message is received. It
+serves as a reply, and also serves to keep the connection alive while
+explicitly notifying the other end that the receiver is still active.
+
+1. type: 19 (`pong`)
+2. data:
+    * [2:byteslen]
+    * [byteslen:randombytes]
+
+#### Requirements
+
+A `pong` message MUST only be sent in response to receiving a `ping` message.
+Within the received `ping` message, the sender will specify the number of bytes
+to be included within the data payload of the `pong` message. Senders of the
+`pong` message may retrieve the payload bytes from any source as the bytes are
+opaque and convey no meaning. Senders of a `pong` message SHOULD take care to
+ensure that the bytes being sent don't contain sensitive data such as secrets,
+or portions of initialized memory.
+
```

Comment:
The non-zero requirement allows for uni-directional pings, which is kind of neat, but do we want that special case in here?

Also with the 30 second min-delta and 65k limit this allows us to continually waste about 2kB/s on the responder's upstream, which may cause problems on some link types. We might want to restrict that later.

We also do not have a timeout period in which the `ping` recipient MUST reply, allowing the sender to eventually close the channel.